### ISSUE TITLE
Auto refresh Active Task field data matrix table.

### ISSUE THREAD

**Issue by bsaiken** [FWAJL/FieldWorkManager#1130]
See Issue #807

**Comment by WebDevJL** [comment]
@bsaiken 

Before suggesting an idea, to extend the functionality, it would be useful, as I think it will happen, to enable the "auto refresh" for all modules in the active task section. Not just the field data or the discussion.

Am I right?

**Comment by bsaiken** [comment]
> @WebDevJL wrote:
Before suggesting an idea, to extend the functionality, it would be useful, as I think it will happen, to enable the "auto refresh" for all modules in the active task section. Not just the field data or the discussion.
Am I right?

We know that the active task discussion module needs to be refreshed.  Now we know we need to refresh the field matrix modules as well.  The active task map has colored markers that show if the location is not started, in process, or finished, and that status could change at any time.  Therefore, I am thinking, rather than auto-refresh individual active task modules, we may want to consider writing the code so that all active tasks modules auto-update.  I opened this Issue to auto-refresh the field data matrix table, but perhaps the scope of the Issue should be all active task modules.  I am asking if you agree if this is a good idea.

**Comment by WebDevJL** [comment]
@bsaiken

Having synced data between pm and ft interfaces is a good idea. To do it or
not is another question. Here are some thoughts.

The first point to take into account is server performance.
Real time auto refresh, or auto refresh every x minutes, can be heavy on
server ressources. Now we have one pm, and maybe 2 at once that could work
on the tool.
Second point is implementation to display up-to-date data.
Using ajax would make the reload of the page unnecessary but data binding
on the current view would be complex to code while reloading the page would
be much more taxing on the server.

With the current application architecture, we need to think what we want to
emphasise : small impact on server ressources or complex implementation.

What do you think?

**Comment by bsaiken** [comment]
@WebDevJL 
I am just not buying this idea that we are putting a big burden on the server by what we are doing.  We have few users and very little data is being passed.  So, I think you may be correct in general, but in this context is it really a significant drain on server resources?  This would only be happening when the PM is logged in, so we can assume that they actually want to see the data.  And, we will be implementing timeout so users do not stay on indefinitely - this is a much for security as for resource management.  Also, we could set it up so it only refreshes the open tab.  Navigating to any new tab would bring up the newest data.

If the updata doesn't bind new data what is the point?  The whole idea of the refresh is so the PM sees the new data without having to refresh the page on their own.  Please clarify if I am misunderstanding the concept of data binding.

Damir already has an auto screen refresh for the communications tab and it seems to work fine, so I do not understand where you can say that the process is complex.  How is this different than the auto-refresh that we use for the communications tab?

Another option we had considered was a refresh button that the PM could click any time they want to make certain that the data displayed is up to date, but with Damir's auto-refresh, we didn't pursue it.

So, give me one more round of feedback and I will make a choice.

Brian

**Comment by FWAJL** [comment]
> @bsaiken said:
> Also, we could set it up so it only refreshes the open tab. Navigating to any new tab would bring up the newest data.

Yes, we can. I was thinking about this as a solution. It limits the requests to the server and we are processing the current data displayed on screen, which is the most important, isn't it?

> @bsaiken said:
> Damir already has an auto screen refresh for the communications tab and it seems to work fine, so I do not understand where you can say that the process is complex. How is this different than the auto-refresh that we use for the communications tab?

Let me correct myself: data-binding would be ***more*** complex than a simple page reload. If done on a tab basis as the PM is viewing its content, then we make less calls to the server to update the views.

The perfect solution is not the goal here and I know that (I know you're thinking about that ;-))
So Damir's auto-refresh would need to be done for the tab(s) where we need it and we would be good to go.

What do you think?

**Comment by bsaiken** [comment]
@WebDevJL 
We will need this done for each tab in the activetask UI.  If you need to add anything to the method implementation, add it here.

**Comment by FWAJL** [comment]
@bsaiken 

Enable an automatic refresh using a similar method that Damir did in the communications tab.

The final specifications are:

1. The web client needs to make AJAX calls to the server to retrieve the data
2. Once the data is retrieved, data-bind it to the view.
3. Make the data-binding reusable since we need to implement it onto all active task tabs.
4. Only new or updated data needs to be retrieved.

Please @WebDevJL or @damirius for questions.

**Comment by bsaiken** [comment]
@WebDevJL 
> Only new or updated data needs to be retrieved.

To be clear.  The data is being viewed by the PM, but it is the FT that is updating the data.  So the update must assume any value could have been changed.
Also, I continue to be perplexed by the limitations and exclusions "Only new or updated data...".  Why not get all of it?  We are not streaming videos, or downloading image files or attachments.  We are only talking about few data bytes here for things like location name and lat/long coordinates, so our bandwidth use is negligible.  Where is this drain on resources you keep alluding to?

**Comment by WebDevJL** [comment]
@bsaiken 

Ok. Let's get everything to make it simple. Comment updated https://github.com/FWAJL/FieldWorkManager/issues/1130#issuecomment-127481599.

**Comment by bsaiken** [comment]
@WebDevJL 
See Issue #1144.  This may be the same Issue.

**Comment by bsaiken** [comment]
@damirius @WebDevJL 
Damir, Please review the comments on this Issue and #1144 and let me know if you are able to implement the solution Jeremie proposes.

**Comment by damirius** [comment]
@bsaiken @WebDevJL 

Discussion auto refresh functions are simple. They are using `setInterval` functions which just run ajax call on set time interval which in turn loads discussion thread. We are saving `lastTimestamp` and send it as a param with that call so that we only fetch new messages instead of refreshing everything. Also there are some "refresh blockers", for example when PM/FT sends a message script will stop refreshing until the message is saved and `lastTimestamp` is updated so that we don't fetch sent message multiple times and so on.

Lots of data is already fetched only through ajax calls. Other stuff that is preloaded are categorized lists mostly, so I guess we want this loaded through ajax too?

Data binding and refreshing. Data binding is a broad term, but basically I guess we want to reflect the changes in the db on any data displayed instantaneously (or close to it). Also we want some set of functions or something similar which is enough flexible that we can reuse it on each part of the active task modules. 

Main reason for refresh is because we want to show most recent data that was edited by someone else or added by someone else. Let's say we have list of locations and we are logged in as FT, PM creates new location and edits old one. Our list should be refreshed and we should see new location as well as changes on the old one. In any case even if we refresh the data we will still read the locations from our session (FT session) and since we didn't edit any of those locations the list will stay the same.
So I guess here comes the change of 'Read-only once from db' to 'Read in consultation in some specific cases'. We still need to see how to catch that specific cases and tell the session to update itself before fetching data from it to display it to the user (FT in this case).

Discussion refresh only works because discussion thread itself isn't stored in the session but it is always fetched from the DB.

**Comment by bsaiken** [comment]
@damirius @WebDevJL @gsouvik 
I am not sure I completely understood your response.  Perhaps a clearer list of constraints and required behaviors will help.

A PM can NOT edit a task that is active.  I think I played around with this, for example, I thought I wanted the PM to be able to edit field data.  If so, my mistake.  Nothing in the Active Task UI should be editable by the PM.  Locations can NOT be added.  The exception to this is sending messages, but that is not editing, it is adding new data.

The PM CAN edit a task that is inactive.  While a Task is inactive, a FT will not be able to access this task, so no conflict.  The conflict arises if a PM wants to deactivate a task and one or more FT is logged in.  Is it possible to identify that one or more FT is logged in, so that we can put a constraint on the task deactivation button?  If so, it's an easy fix.  Instead of deactivating, we have an alert box that notifes the PM that [technician_name] is currently logged in.  Then the PM could send them a message telling them they need to log out.

The PM must be able to see the updates as the FT enters new data, including adding a new location.  This does not need to be "instantaneous", but on a reasonable update interval (say, 1 minute).  Right now, the PM would need to logout to see new data, which is unacceptable.  I think what Jeremie means by "specific cases" is certain view modules.  So, all Active Task view modules (tabs), and everything displayed on those pages, would qualify as a "special case".  Please clarify for me why we see new messages without logging back in, but we can't see a newly added Location.

**Comment by damirius** [comment]
@bsaiken @WebDevJL 

First half of the comment is meant for Jeremie, so we can find the way to implement those changes in the code.
In any case I gave wrong example, but the point is the same. If you are PM you can't see the changes FT makes to the locations even with refreshing because refreshing will fetch the data from sessions and not from the db.

Last paragraph in my previous comment explains why you can see messages:
>Discussion refresh only works because discussion thread itself isn't stored in the session but it is always fetched from the DB.

So you can see new messages because I didn't want to use sessions to store discussion thread there, instead we are always loading NEW messages from the db.

**Comment by bsaiken** [comment]
@damirius 
Your example was adequate.  I understand that the answer is to pull data from the DB, not the session array, or change the way the session array is updated, so it updates with a page refresh.  

I have been opposed to the session arrays from the beginning, in part because I could foresee issues like this occurring.  However, part of my resistance has been due to a lack of sufficient knowledge on web development strategies.    

Jeremie suggested implementing a "read in consultation" approach (https://github.com/FWAJL/FieldWorkManager/issues/1144#issuecomment-127899127).  I'm not certain what this means.  Do you understand this terminology, and would you be able to implement the approach?

**Comment by WebDevJL** [comment]
@bsaiken 

First, thank you for laying down clearly what are your needs Brian. I think it clarifies it for everybody

The use of the session can be skipped if you're really against it and we can look at performances later. It'll save us time now.
However, I want to emphasize that, if implemented properly, e.g. you have for the session data methods to read, remove and add data into it, we wouldn't have this "problem". We did not take the time to do it fully for every new data that was stored using the website as we progressed in building the website. 
So now, in order to move on and accomplish the tasks at hand, we better just read/write from/to the DB without the session in this case of Active Task.

If we need to do it for other cases, let's just forget the session and do the dev work to build the functionalities the website needs. It will be a debate when we actually need to worry about performance.

Pardon me my bad english : by "read in consultation", I meant querying the database to select some data. Like you described clearly, PM doesn't modify active task data, he only looks at it.

Can we now start the dev work or is there any point left in the dark?

**Comment by WebDevJL** [comment]
@bsaiken 

I suggest to have one issue per view where we want the auto refresh. This issue will be the "documentation" issue.

Is that ok with you?

**Comment by bsaiken** [comment]
@damirius @gsouvik 
Please review new Issue assignments dealing with this Issue.  Where an Issue became a discussion, like this one, I have labeled it "Milestone info" and assigned all related Issues to the Milestone "Active Task UI".
If you have any questions let me know.
My goal is to get the Medium, High, and Very High priority Issues done be August 17.
